### ISSUE TITLE
Extraction error when using useTranslation hook and passing t function as a parameter to another function

### ISSUE THREAD

**Issue by bopfer** [gilbsgilbs/babel-plugin-i18next-extract#69]
To get the `t` function, all my components have this:

```javascript
const { t } = useTranslation();
```

When I extract translations, I get an error like this for each file that uses the hook:

```
babel-plugin-i18next-extract: Extraction error in <filename> at line xxx. Couldn't
evaluate i18next key. You should either make the key evaluable or skip the line
using a skip comment (/* i18next-extract-disable-line */ or
/* i18next-extract-disable-next-line */).
```
> Where `line xxx` is the line shown above to get access to `t`.

Is there a way to have the extractor avoid this line without having to add the comment hints?

**Comment by gilbsgilbs** [comment]
Hi! I don't think the error is at the line you pointed out. Can you try to find which usage of the `t` function generate this error message? It is true that the error message isn't very correct though.

**Comment by bopfer** [comment]
It definitely is that line.  Here is an exact error for a specific file:

```
babel-plugin-i18next-extract: Extraction error in
src/features/Admin/pages/UserProfile/ProfileForm.tsx at line 104.
Couldn't evaluate i18next key. You should either make the key evaluable
or skip the line using a skip comment (/* i18next-extract-disable-line */ or
/* i18next-extract-disable-next-line */).
```

And the code at that line:

![Screen Shot 2019-09-23 at 1 36 18 PM](https://user-images.githubusercontent.com/824368/65460744-70caff80-de07-11e9-972b-c32242e24b70.png)

Interestingly, the comment hints also don't work to stop the error.

**Comment by gilbsgilbs** [comment]
What I'm trying to say is that even though the log indicates that the line 104 is at fault, it is probably **not** the case and it's likely to be a bug somewhere else. It must be one usage of the `t` function somewhere after line 104 that generates the warning.

The fact that it indicates the wrong line is another bug.

**Comment by bopfer** [comment]
Ahh, could be that.  I do pass the `t` function into other functions as an argument.  Looks like that is causing it.  The error goes away if I put the comment before each of those function calls.

For example, I have stuff like this:

```javascript
const errorMap = getErrorMap(t);
```

The error goes away if I do this:

```javascript
/* i18next-extract-disable-next-line */
const errorMap = getErrorMap(t);
```

**Comment by gilbsgilbs** [comment]
Yes this is it. Passing `t` function as a parameter to other functions is definitely not supported and won't work as expected. If you happen to call your parameter `t` in `getErrorMap` function, keys would still be extracted in there, but `useTranslation`-specific features such as namespaces wouldn't work properly.

After having a closer look at the plugin code, the fact that this specific warning shows is however unintended and quite unlucky. I'm not sure what to do but given the current state of the code, it seems that the initial intent was indeed to not show anything in such cases. Maybe we should stick to it.

By the way [this PR](https://github.com/gilbsgilbs/babel-plugin-i18next-extract/pull/70) should fix the line numbers showing up in warnings.

**Comment by bopfer** [comment]
Yeah, the extracting from the other functions is working properly. I am using the comment hints to set the namespace for each file.  So, the `useTranslation`-specific features aren't a big deal.  They would be a nice thing to have.  Not sure how possible that even is though.

**Comment by gilbsgilbs** [comment]
Fixed in 0.2.6 :tada: .

> So, the useTranslation-specific features aren't a big deal. […] Not sure how possible that even is though.

I can see at least two major reasons why it's very unlikely we'll ever be able to handle such cases using static analysis:

- When passing `t` as a parameter to a function, all the context you had in it is lost. You may call the same function with another instance of `t` somewhere else and therefore keeping track of which namespace is bound to which instance of the `t` function would become a nightmare.
- As far as I know, Babel just parses one file after another and doesn't create links between the symbols of files (like an IDE would do). That's why keeping contexts across functions that may be imported from another file would be even harder if not impossible using Babel.
